dartagnan/src/main/java/com/dat3m/dartagnan/program/Function.java

```diff
@@ -0,0 +1,118 @@
+package com.dat3m.dartagnan.program;
+
+import com.dat3m.dartagnan.exception.MalformedProgramException;
+import com.dat3m.dartagnan.expression.type.FunctionType;
+import com.dat3m.dartagnan.expression.type.Type;
+import com.dat3m.dartagnan.program.event.core.Event;
+import com.google.common.base.Preconditions;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class Function {
+
+    protected String name;
+    protected Event entry; // Can be null for intrinsics
+    protected int id;
+    protected Event exit;
+
+    protected FunctionType functionType;
+    protected List<Register> parameterRegs;
+
+    protected Program program;
+    protected Map<String, Register> registers = new HashMap<>();
+    protected int dummyCount = 0;
+
+    public Function(String name, FunctionType type, List<String> parameterNames, int id, Event entry) {
+        Preconditions.checkArgument(type.getParameterTypes().size() == parameterNames.size());
+        this.name = name;
+        this.functionType = type;
+        this.id = id;
+        this.entry = entry;
+
+        parameterRegs = new ArrayList<>(parameterNames.size());
+        int paramIndex = 0;
+        for (Type paramType : functionType.getParameterTypes()) {
+            final Register paramReg = newRegister(parameterNames.get(paramIndex++), paramType);
+            parameterRegs.add(paramReg);
+        }
+
+        Event cur = entry;
+        while (cur != null) {
+            cur.setFunction(this);
+            exit = cur;
```

Comment:
do we really need to update in every iteration? why not simply at the end?

Reply:
You mean `exit`? At the end of the loop `cur == null`, so you do not know what the exit event was, unless you track it, which is exactly what the code does.
If you talk about `setFunction`, then this needs to get executed for every event of course. 
The whole loop is just to make sure that if you create a function with an entry event that already has successors, then all the successors are added to the function as well.
That being said, I should probably make sure that `entry.getPredecessor() == null` holds.

Reply:
I was talking about `exit` so your first comment answers my question already